### PR TITLE
🍪`react-cookie` package

### PR DESCRIPTION
**Pull request by cartogram**
## Description

~~This PR adds a `@shopify/react-cookie` library that gathers cookies from both the server and client. There are three main components to this library, 1) A universal-provider that _provides_ a set of initial cookies and instantiates 2) a manager. This manager provides common methods for getting/ setting cookies both internally, and in the `document`. 3) Finally, there are 2 common React hooks for operating on the cookies from within a React component.~~

#### Demo/ Tophat
You can view a [demo here](https://monosnap.com/file/YJuB4TUw6etygwhCj5A4DA38r1Dlmt) from [this branch](https://github.com/Shopify/web-rails-proving-ground/compare/cookies?expand=1) of @shopify/web-rails-proving-ground

If you want to tophat yourself, pull down this branch and also the [`cookies`]() branch from `web-rails-proving-ground`.

1. `dev up` in the `cookies` branch of `web-rails-proving-ground`
1. From the quilt repo run:
```
cp -R packages/react-cookie/src/ ../web-rails-proving-ground/packages/\@shopify/react-cookie
cp -R packages/react-network/src/ ../web-rails-proving-ground/packages/\@shopify/react-network
cp -R packages/react-server/src/ ../web-rails-proving-ground/packages/\@shopify/react-server
```
1. Then run `dev run` in the `web-rails-proving-ground` repo.

### CODE REVIEW

File: packages/react-cookie/src/hooks.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+export function useCookie(
+  key: string,
+): [
+  string | undefined,
+  (value?: string, options?: CookieSerializeOptions) => void
+] {
+  const manager = useContext(CookieContext);
+
+  if (!manager) {
+    throw new Error(NO_MANAGER_ERROR);
```
Comment (cartogram) [inline packages/react-cookie/src/hooks.ts]: Does it make sense to `throw` here?

File: packages/react-cookie/README.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
+# `@shopify/react-cookie`
```
Comment (cartogram) [inline packages/react-cookie/README.md]: Just roughed this out for completeness, but will put some more ❤️  into this `Readme` either before or immediately after merging.

File: packages/react-cookie/src/tests/hooks.test.tsx [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+      const key = 'foo';
+      const value = 'bar';
+      const cookies = {[key]: value};
+
+      const wrapper = mount(<MockComponent cookie={key} />, {
+        manager: new CookieManager({...cookies}),
+      });
+
+      wrapper
+        .find(MockComponent)!
```
Comment (cartogram) [inline packages/react-cookie/src/tests/hooks.test.tsx]: This `find` returns `null` occasionally and is why CI is failing. I am not sure why yet.

File: packages/react-cookie/package.json [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  },
+  "bugs": {
+    "url": "https://github.com/shopify/quilt/issues"
+  },
+  "homepage": "https://github.com/Shopify/quilt/blob/master/packages/react-cookie/README.md",
+  "dependencies": {
+    "@shopify/react-effect": "^3.2.3",
+    "@shopify/react-hooks": "^1.2.1",
+    "@shopify/react-html": "^9.0.5",
+    "cookie": "^0.4.0",
```
Comment (cartogram) [inline packages/react-cookie/package.json]: TODO: `@shopify/react-network` is missing here.

File: packages/react-cookie/src/hooks.ts (hunk 1 above)
Comment (TheMallen) [reply packages/react-cookie/src/hooks.ts]: I think so?

File: packages/react-cookie/src/hooks.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+  }
+
+  const [cookie, setCookieValue] = useState(
+    () => (manager ? manager.getCookie(key) : undefined),
+  );
+
+  const setCookie = (value?: string, options?: CookieSerializeOptions) => {
+    if (!manager) {
```
Comment (TheMallen) [inline packages/react-cookie/src/hooks.ts]: Manager should always be defined given that we are throwing before we construct this when it isn't.

This is actually a great usecase for the upcoming [assert signatures](https://httptoolkit.tech/blog/5-big-features-of-typescript-3.7/) feature, but till we have that I'd just `manager!` or cast it.

Comment (TheMallen) [reply packages/react-cookie/src/hooks.ts]: on a side note it feels kinda weird that this lives here and not in the manager itself.

File: packages/react-cookie/src/manager.ts [hunk 6]
```diff
@@ -0,0 +1,65 @@
+import cookie from 'cookie';
+import {hasDocumentCookie} from './utilities';
+import {Cookies, CookieValue} from './types';
+
+export {CookieContext} from './context';
+
+export class CookieManager {
```
Comment (TheMallen) [inline packages/react-cookie/src/manager.ts]: tbh I'm not super sure why we aren't using `universal-cookies` as our manager, does it clash with our expected APIs in some way? It's simple enough that it's not a huge deal for us to have our own but if there's no real reason not to we might as well use it.

**Comment by cartogram** [comment]
@TheMallen on doing our own thing vs third-party: this library allows us to provide a cookie solution for applications devs that uses consistent patters as with the other quilt packages that also require similar client/server interaction (self-serializer, react-effect, network, etc…) . Going this route also requires no server-side middleware and everything is confined to only this package (and really one component) rather than requiring orchestrated changes across `@shopify/react-server`, `@shopify/sewing-kit`, etc. 

Less important, there is only one viable third-party solution out there (the `universal-cookie`/ `react-cookie` stack) and I wasn’t the biggest fan of the imperative change-listener way that those libraries operate. I wanted to explore a way that we could leverage our own libraries to provide something better.

That said, I get that additional code is a liability (even given this library is pretty tiny) and understand the hesitation to introduce this package.

File: packages/react-cookie/src/Cookie.tsx [hunk 7]
```diff
@@ -0,0 +1,32 @@
+import React from 'react';
```
Comment (GoodForOneFare) [inline packages/react-cookie/src/Cookie.tsx]: This follows the universal provider pattern so closely, would it make sense to call this `CookiesUniversalProvider`?

File: packages/react-cookie/src/Cookie.tsx [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+    return new CookieManager(initialCookies || cookies);
+  }).current;
+
+  return (
+    <>
+      <CookieContext.Provider value={manager}>
+        {children}
+      </CookieContext.Provider>
+      <Serialize data={() => manager.getCookies()} />
```
Comment (GoodForOneFare) [inline packages/react-cookie/src/Cookie.tsx]: I don't understand this 😕 Why wouldn't cookies be sent as response headers?

Comment (cartogram) [reply packages/react-cookie/src/Cookie.tsx]: The `useRequestHeader ` hook returns `undefined` on the client so without this, the server cookies are lost.

Comment (GoodForOneFare) [reply packages/react-cookie/src/Cookie.tsx]: I haven't worked much with cookies, but as I understand it, this should be possible:
- The server side of this generates `Set-Cookie` headers for each name/value pair
- The browser sees these headers, and populates `document.cookie` to reflect them
- The client side of this grabs `document.cookie`, and parses it into name/value pairs

Not an expert on this, so I could definitely be wrong.

Comment (TheMallen) [reply packages/react-cookie/src/Cookie.tsx]: The above seems right

File: packages/react-cookie/src/CookieUniversalProvider.tsx [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+    return new CookieManager(initialCookies || cookies);
+  }).current;
+
+  return (
+    <>
+      <CookieContext.Provider value={manager}>
+        {children}
+      </CookieContext.Provider>
+      <Serialize data={() => manager.getCookies()} />
```
Comment (TheMallen) [inline packages/react-cookie/src/CookieUniversalProvider.tsx]: as Gord mentioned, we shouldn't need to actually serialize the cookies since we can build the manager out of `document.cookie` on the client anyway.

File: packages/react-cookie/src/CookieUniversalProvider.tsx [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+}
+
+export function CookieUniversalProvider({children}: Props) {
+  const [initialCookies, Serialize] = useSerialized<Cookies>('cookies');
+  const serverCookieHeader = useRequestHeader('cookie');
+
+  const manager = useLazyRef(() => {
+    const cookies = hasDocumentCookie() ? document.cookie : serverCookieHeader;
+
+    return new CookieManager(initialCookies || cookies);
```
Comment (TheMallen) [inline packages/react-cookie/src/CookieUniversalProvider.tsx]: `new CookieManager(cookies)` should be fine here

File: packages/react-cookie/src/manager.ts [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    for (const [key, value] of this.cookies) {
+      cookies[key] = value;
+    }
+
+    return {
+      ...cookies,
+    };
+  }
+
+  setCookie(name: string, value: string | CookieValue) {
```
Comment (TheMallen) [inline packages/react-cookie/src/manager.ts]: I think realistically on the server we want to set the cookie using a `networkManager` form `@shopify/react-network`,  which would probable entail

- depending on it here and passing it in from the component, or rolling this back into that library
- adding a `setCookie` property on the network manager
- adding logic to `applyToContext` that calls `ctx.cookies.set` (which will handle setting `set-cookie` as a user would expect)

File: packages/react-cookie/src/UniversalCookies.ts [hunk 12]
```diff
@@ -0,0 +1,54 @@
+import cookie from 'cookie';
+import {Context} from 'koa'
+import {Cookie, CookieValue} from './types';
+
+export {CookieContext} from './context';
+
+export class UniversalCookies {
```
Comment (cartogram) [inline packages/react-cookie/src/UniversalCookies.ts]: open to better names here

File: packages/react-cookie/src/UniversalCookies.ts [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+    const fullCookie = typeof value === 'string' ? {value} : value;
+
+    this.cookies.set(name, fullCookie);
+  }
+
+  removeCookie(name: string) {
+    this.cookies.delete(name);
+  }
+
+  applyToContext(ctx: Context) {
```
Comment (cartogram) [inline packages/react-cookie/src/UniversalCookies.ts]: embedding the server logic into the class

File: packages/react-cookie/src/context.ts [hunk 14]
```diff
@@ -0,0 +1,7 @@
+import React from 'react';
+import {CookieManager} from './manager';
+import {UniversalCookies} from './UniversalCookies';
+
+export const CookieContext = React.createContext<
+  CookieManager | UniversalCookies | null | any
```
Comment (cartogram) [inline packages/react-cookie/src/context.ts]: todo: remove any

File: packages/react-cookie/README.md [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+To use the hooks provided by this library you must first wrap your application tree in the `<CookieUniversalProvider />` component.
+
+```tsx
+// App.tsx
+
+import {CookieUniversalProvider} from '@shopify/react-cookie';
+import {SomeComponent} from './someComponent';
+
+function App() {
+  <Cookie>
```
Comment (TheMallen) [inline packages/react-cookie/README.md]: `CookieUniversalProvider`

File: packages/react-cookie/src/CookieUniversalProvider.tsx [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+
+export interface Props {
+  children?: React.ReactNode;
+}
+
+export function CookieUniversalProvider({children}: Props) {
+  const manager = useNetworkManager();
+
+  const cookieManager = useLazyRef(() => {
+    return !hasDocumentCookie() && manager
```
Comment (TheMallen) [inline packages/react-cookie/src/CookieUniversalProvider.tsx]: this seems slightly wrong to me since it could make a front-end focused cookie manager if the user forgot to wrap their tree in a network provider or it isn't there for some other weird reason. I'd rather we do something like

```tsx
if (hasDocumentCookie()) {
  return new CookieManager();
}

if (manager == null) {
  throw new error('You fool! You fell victim to one of the classic blunders! Never server render a react-html app without a network provider when death is on the line!');
}

return manager.cookies;
```

File: packages/react-cookie/src/context.ts (hunk 14 above)
Comment (TheMallen) [reply packages/react-cookie/src/context.ts]: IMO we just want to have an `interface CookieManager {}` that lays out the API that both of these managers use.

File: packages/react-cookie/src/UniversalCookies.ts (hunk 12 above)
Comment (TheMallen) [reply packages/react-cookie/src/UniversalCookies.ts]: This one only works on the server right? ServerCookieManager?

File: packages/react-cookie/src/UniversalCookies.ts (hunk 13 above)
Comment (TheMallen) [reply packages/react-cookie/src/UniversalCookies.ts]: 🤔 seems weird to me, I'd expect it to be like `react-network` and expose a function separately (that way the manager isn't tied to koa, just the `applyToContext` function is)

```
import {ServerCookieManager} '@shopify/react-cookie'
import {applyToContext} from '@shopify/react-cookie/koa'

// in the middleware
applyToContext(ctx, cookies);
```

File: packages/react-cookie/src/tests/cookie.ts [hunk 17]
```diff
@@ -0,0 +1,15 @@
+export default class Cookie {
+  mock(value: string) {
```
Comment (TheMallen) [inline packages/react-cookie/src/tests/cookie.ts]: should this go live in `jest-dom-mocks`?

File: packages/react-network/src/manager.ts [hunk 18]
```diff
@@ -1,12 +1,14 @@
 import {StatusCode, CspDirective, Header} from '@shopify/network';
 import {EffectKind} from '@shopify/react-effect';
+import {Cookie, UniversalCookies} from '@shopify/react-cookie';
```
Comment (TheMallen) [inline packages/react-network/src/manager.ts]: isn't this a circular dependency? Should the server cookie manager live in `@shopify/react-network`?

File: packages/react-cookie/src/context.ts (hunk 14 above)
Comment (cartogram) [reply packages/react-cookie/src/context.ts]: Yeah that works

File: packages/react-cookie/src/tests/cookie.ts (hunk 17 above)
Comment (cartogram) [reply packages/react-cookie/src/tests/cookie.ts]: yeah it should 👍

File: packages/react-network/src/manager.ts (hunk 18 above)
Comment (cartogram) [reply packages/react-network/src/manager.ts]: Yeah also agree with this

File: packages/react-cookie/src/tests/cookie.ts (hunk 17 above)
Comment (cartogram) [reply packages/react-cookie/src/tests/cookie.ts]: I'm going to do this in a follow up though. There are already a lot of packages involved in this PR and this is an easy thing to extract afterward.

**Comment by cartogram** [comment]
@TheMallen another look when you get a chance 🙇 Still doing a final comb through all the Readme, but otherwise all changes are in.

File: packages/react-cookie/README.md [hunk 19]
```diff
@@ -0,0 +1,132 @@
...
+
+function App() {
+  <CookieUniversalProvider>
+    // rest of your tree
+    <SomeComponent />
+  </CookieUniversalProvider>;
+}
+```
+
+You can then use the [hooks](#hooks) provided by this package. See [hooks](#hooks) below for more information.
```
Comment (TheMallen) [inline packages/react-cookie/README.md]: I don't think you need the second sentence here since you already have a link.

File: packages/react-cookie/README.md [hunk 20]
```diff
@@ -0,0 +1,132 @@
...
+
+```bash
+$ yarn add @shopify/react-cookie
+```
+
+## Usage
+
+### Server
+
+To extract cookies during the server-side render of your application, your component need to have access to the `NetworkManager` from `@shopify/react-network`. You can pass the initial server cookie value when the manager is instantiated within your server side middleware.
```
Comment (TheMallen) [inline packages/react-cookie/README.md]: needs*

File: packages/react-cookie/src/server.ts [hunk 21]
```diff
@@ -0,0 +1,12 @@
+import {Context} from 'koa';
```
Comment (TheMallen) [inline packages/react-cookie/src/server.ts]: I still think the way these two packages are broken up is strange. This function in particular is strange when the rest of the server cookie stuff is just part of react-network. Why can't this just be part of `react-network`'s `applyToContext`?

Comment (cartogram) [reply packages/react-cookie/src/server.ts]: It can be 👍 Slightly related, I am considering pulling the shared types and server manager into a more generic `@shopify/cookie` package that `react-network` and `react-cookie` would depend on.

File: packages/react-cookie/src/server.ts [hunk 22]
```diff
@@ -0,0 +1,12 @@
...
+import {Cookie} from './types';
+
+export function applyCookieToContext(cookies: Cookie) {
+  return function(ctx: Context) {
+    Object.entries(cookies).forEach(([cookie, options]) => {
+      const {value, ...cookieOptions} = options;
+
+      ctx.cookies.set(cookie, value, cookieOptions as any);
```
Comment (GoodForOneFare) [inline packages/react-cookie/src/server.ts]: Could you add `none` in a PR to https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/cookies/index.d.ts#L91, please?

File: packages/react-network/src/ServerCookieManager.ts [hunk 23]
```diff
@@ -0,0 +1,43 @@
+import cookie, {CookieSerializeOptions} from 'cookie';
```
Comment (GoodForOneFare) [inline packages/react-network/src/ServerCookieManager.ts]: Should this class have a unit test?

File: packages/react-cookie/src/BrowserCookieManager.ts [hunk 24]
```diff
@@ -0,0 +1,65 @@
+import cookie from 'cookie';
+import {hasDocumentCookie} from './utilities';
+import {CookieValue, Cookie} from './types';
+
+export {CookieContext} from './context';
+
+export class BrowserCookieManager {
```
Comment (GoodForOneFare) [inline packages/react-cookie/src/BrowserCookieManager.ts]: I guess this is consistent with the rest of Quilt, but I miss the lint rule for "this isn't really a class".

File: .vscode/settings.json [hunk 25]
```diff
@@ -10,7 +10,6 @@
     "**/.git": true,
     "**/.svn": true,
     "**/.hg": true,
-    "**/node_modules": true,
```
Comment (GoodForOneFare) [inline .vscode/settings.json]: Please revert this change before merging.

File: packages/react-cookie/server.d.ts [hunk 26]
```diff
@@ -0,0 +1 @@
+export * from './dist/server';
```
Comment (GoodForOneFare) [inline packages/react-cookie/server.d.ts]: I don't think you can export useful types from a js file.

File: packages/react-cookie/README.md [hunk 27]
```diff
@@ -0,0 +1,132 @@
...
+      <p>the cookie is {cookie}</p>
+      <TextField value={cookie} onChange={handleChange} />
+    </>
+  );
+}
+```
+
+### Hooks
+
+#### `useCookies()`
```
[the remark this answers is not in the document]
Comment (cartogram) [reply packages/react-cookie/README.md]: It was helpful for testing, but I can't imagine purpose for consumers so I am removing it for now. 👍

File: packages/react-cookie/src/BrowserCookieManager.ts [hunk 28]
```diff
@@ -0,0 +1,65 @@
...
+  getCookie(name: string) {
+    if (!hasDocumentCookie()) {
```
[the remark this answers is not in the document]
Comment (cartogram) [reply packages/react-cookie/src/BrowserCookieManager.ts]: Good catch, was left in from a "universal" implementation of this manager.

File: packages/react-cookie/src/hooks.ts [hunk 29]
```diff
@@ -0,0 +1,52 @@
...
+      manager.removeCookie(key);
+
+      return;
+    }
+
+    setCookieValue(value);
+    manager.setCookie(key, {value, ...options});
+  };
+
+  return [cookie, setCookie];
```
[the remark this answers is not in the document]
Comment (cartogram) [reply packages/react-cookie/src/hooks.ts]: This would break our lint rule https://github.com/Shopify/eslint-plugin-shopify/blob/master/docs/rules/react-hooks-strict-return.md. 

I opted for "setting" to null/undefined to remove the cookie, but I did not add documentation or tests for this feature. Those are both added now and I think provides a simple API for consumers. Do you think an explicit `removeCookie()` function is needed?

**Comment by cartogram** [comment]
@TheMallen @GoodForOneFare All changes have been addressed. I also included:

- A `react-server` test that sending cookies with the request will be accessible via the client-side hooks.
- Added the UniversalCookieProvider to our default providers
- Tests for ServerCookieProvider 
- Tests for the react-network/server component
- More docs for all the changed things
- Utils for creating and clearing cookies in tests
- probably much more that I am forgetting

The only way I was able to tophat this successfully was by building each library (`react-server`, `react-cookie` and `react-network`) and copying them into web-rails-proving-ground`'s `node_modules. There is a branch there called `cookies` that has some application code for testing the hooks in this package.

Thanks again for your continual feedback and patience!

File: packages/react-cookie/src/BrowserCookieManager.ts (hunk 24 above)
Comment (TheMallen) [reply packages/react-cookie/src/BrowserCookieManager.ts]: Yeah this could easily just be a constant object exported directly since it doesn't have any local state. Doesn't matter too much.

File: packages/react-cookie/src/hooks.ts (hunk 29 above)
[the remark this answers is not in the document]
Comment (TheMallen) [reply packages/react-cookie/src/hooks.ts]: Personally I like the consistency with koa of doing `set('name', null)` as long as it's documented

File: packages/react-cookie/src/tests/CookieUniversalProvider.test.tsx [hunk 30]
```diff
@@ -0,0 +1,62 @@
...
+  afterEach(() => {
+    hasDocumentCookie.mockClear();
+  });
+
+  function MockApp() {
+    return null;
+  }
+
+  it('provides a universal cookie manager for the server', () => {
+    hasDocumentCookie.mockImplementation(false);
```
Comment (TheMallen) [inline packages/react-cookie/src/tests/CookieUniversalProvider.test.tsx]: we might want to add a 'why' comment explaining why we have to mock this (so that we can test the server implementation despite jest always having a document.cookie defined)

File: packages/react-cookie/src/tests/hooks.test.tsx [hunk 31]
```diff
@@ -0,0 +1,115 @@
...
+    });
+
+    it('removes a cookie', async () => {
+      const setValue = 'bar';
+
+      const wrapper = await mount(
+        <MockComponent cookie="foo" setValue={setValue} />,
+      );
+
+      wrapper.find('button')!.trigger('onClick');
```
Comment (TheMallen) [inline packages/react-cookie/src/tests/hooks.test.tsx]: this spacing is a bit weird IMO

File: packages/react-cookie/README.md [hunk 32]
```diff
@@ -0,0 +1,110 @@
...
+    <>
+      <p>the cookie is {cookie}</p>
+      <TextField value={cookie} onChange={handleChange} />
+      <button type="button" onClick={removeCookie}>
+        Remove Cookie
+      </button>
+    </>
+  );
+}
+```
```
Comment (TheMallen) [inline packages/react-cookie/README.md]: Can we also document the test utilities we export?

File: packages/react-cookie/src/hooks.ts (hunk 29 above)
[the remark this answers is not in the document]
Comment (GoodForOneFare) [reply packages/react-cookie/src/hooks.ts]: If it makes sense to follow our rule, I guess it's fine.  I'm not 100% sure about how strict we about following it.

Is there any kind of type-hint we can provide like:
```
export const UNSET = '';
...
setCookie(name: string, value: string | CookieValue | UNSET)
```

Eww :/
